crates/starlight/src/vm/promise.rs

```diff
@@ -0,0 +1,301 @@
+use super::arguments::*;
+use super::method_table::*;
+use super::object::*;
+use super::string::*;
+use super::symbol_table::Internable;
+use super::value::*;
+use super::Runtime;
+use crate::gc::cell::{Trace, Tracer};
+use crate::gc::snapshot::deserializer::Deserializer;
+use crate::gc::snapshot::serializer::SnapshotSerializer;
+use crate::vm::function::JsClosureFunction;
+use crate::vm::structure::Structure;
+use std::mem::ManuallyDrop;
+
+pub struct JsPromise {
+    subs: Vec<(Option<JsValue>, Option<JsValue>, Option<JsValue>, JsValue)>, // then_func / catch_func / finally_func / sub_promise
+    resolution: Option<Result<JsValue, JsValue>>,
+}
+
+#[allow(non_snake_case)]
+impl JsPromise {
+    pub fn new(vm: &mut Runtime, function_value: JsValue) -> Result<JsValue, JsValue> {
+        let promise = Self::new_unresolving(vm)?;
+
+        // add persistentrooted for promise and function_value here..
+        let prom_rooted_id = vm.add_persistent_root(promise);
+        let func_rooted_id = vm.add_persistent_root(function_value);
+
+        let sched_res = vm.schedule_async(move |vm| {
```

Comment:
I think `schedule_async` should be wrapped in `resolve_func` and `reject_func`. Because `function_value` should be invoked immediately

Reply:
Thanks for the comment!
That does seem to be what browsers do, allthough i find this surprising. 
see also https://stackoverflow.com/questions/42118900/when-is-the-body-of-a-promise-executed
i'll fix this when i handle the other todos (like calling subs/then/catch async)